Remove base DAM mismatch warning on types

sbomer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/118713 ~(analyzer only for now)~
Fixes https://github.com/dotnet/runtime/issues/119290

This removes the warning for DAM annotation mismatch in the base types of a type declaration. This doesn't cause holes because we produce warnings on the construction of the derived type. Removing the warning enables the use of static methods on such a type.

While implementing this I ran into an ILLink bug around branch removal. We weren't fixing up IL offsets during branch removal, but we were using the offsets as a dictionary key when tracking known stacks during method body scanning. This includes a change to fix up IL offsets during the IL rewrite when we remove unreachable blocks.

Depends on https://github.com/dotnet/runtime/pull/119291

sbomer committed:
Fix new constraint warning

sbomer committed:
Avoid duplicate warning, add test

sbomer committed:
Move arguments to instance fields in GenericArgumentDataFlow

sbomer committed:
Remove analyzer DAM warning on types

sbomer committed:
Add back interface warning

Moving this warning is more complicated than that for base types
because we would either need to warn on conversion to interfaces,
or warn on every constructor (for kept interface implementations).

sbomer committed:
Remove base warning for ILLink

sbomer committed:
Same for ILC

sbomer committed:
Fix ILLink generic behavior

- Move new constraint processing to generic dataflow
- Take new constraint into account for requires dataflow check

sbomer committed:
PR feedback

sbomer committed:
Merge branch 'newConstraintHole' into removeBaseDAMWarning

sbomer committed:
Add test to analyzer tests

sbomer committed:
Fix offsets when rewriting IL

MethodBodyScanner tracks known stacks by IL offset,
so these need to be correct when analyzing IL. We can't just
wait for cecil to update offsets for us when writing to disk.

sbomer committed:
Merge remote-tracking branch 'origin/main' into removeBaseDAMWarning

jtschuster reviewed (approved):
It feels a bit awkward to be passing around the instances of the analyzers, but I don't have a better solution.

sbomer committed:
PR feedback

- Use interface for generic dataflow
- Remove outdated comment
- Clarify expected warning reason
- Add Kept attributes for clarity

jtschuster committed:
Merge branch 'main' into removeBaseDAMWarning
